File: src/main/java/org/strassburger/lifestealz/listeners/WorldSwitchListener.java
package org.strassburger.lifestealz.listeners;

import org.bukkit.World;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerChangedWorldEvent;
import org.strassburger.lifestealz.LifeStealZ;
import org.strassburger.lifestealz.util.storage.PlayerData;

import java.util.List;

public class WorldSwitchListener implements Listener {
    private final LifeStealZ plugin;

    public WorldSwitchListener(LifeStealZ plugin) {
        this.plugin = plugin;
    }

    @EventHandler
    public void onWorldSwitch(PlayerChangedWorldEvent event) {
        Player player = event.getPlayer();
        World fromWorld = event.getFrom();
        World toWorld = player.getLocation().getWorld();

        List<String> worldWhitelist = plugin.getConfig().getStringList("worlds");

        if (worldWhitelist.contains(toWorld.getName())) {
            handleWhitelistedWorld(player, fromWorld, worldWhitelist);
        } else {
            LifeStealZ.setMaxHealth(player, 20.0);
        }
    }

    private void handleWhitelistedWorld(Player player, World fromWorld, List<String> worldWhitelist) {
        PlayerData playerData = plugin.getStorage().load(player.getUniqueId());
        LifeStealZ.setMaxHealth(player, playerData.getMaxhp());

        if (!worldWhitelist.contains(fromWorld.getName())) {
            player.setHealth(playerData.getMaxhp());
        }
    }
}

File: src/main/java/org/strassburger/lifestealz/listener/WorldSwitchListener.kt
package org.strassburger.lifestealz.listener

import org.bukkit.attribute.Attribute
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerChangedWorldEvent
import org.strassburger.lifestealz.Lifestealz
import org.strassburger.lifestealz.util.ManagePlayerdata

class WorldSwitchListener : Listener {
    @EventHandler
    fun worldSwitchFunction(event: PlayerChangedWorldEvent) {
        val player = event.player

        val worldWhitelist = Lifestealz.instance.config.getList("worlds")

        if (worldWhitelist == null) return

        if (worldWhitelist.contains(player.location.world.name)) {
            val playerData = ManagePlayerdata().getPlayerData(name = player.name, uuid = player.uniqueId.toString())
            setMaxHealth(player, playerData.maxhp)

            if (!worldWhitelist.contains(event.from.name)) player.health = playerData.maxhp
        } else {
            setMaxHealth(player, 20.0)
        }
    }

    private fun setMaxHealth(player: Player, maxHealth: Double) {
        val attribute = player.getAttribute(Attribute.GENERIC_MAX_HEALTH)
        if (attribute != null) {
            attribute.baseValue = maxHealth
        }
    }
}